integrations/langchain-js/src/BraintrustCallbackHandler.ts

```diff
@@ -0,0 +1,617 @@
+import { ExperimentLogPartialArgs, isObject } from "@braintrust/core";
+import {
+  BaseCallbackHandler,
+  BaseCallbackHandlerInput,
+  HandleLLMNewTokenCallbackFields,
+  NewTokenIndices,
+} from "@langchain/core/callbacks/base";
+import { AgentAction, AgentFinish } from "@langchain/core/dist/agents";
+import { DocumentInterface } from "@langchain/core/dist/documents/document";
+import { Serialized } from "@langchain/core/dist/load/serializable";
+import { BaseMessage } from "@langchain/core/dist/messages/base";
+import {
+  ChatGeneration,
+  ChatResult,
+  Generation,
+  LLMResult,
+} from "@langchain/core/dist/outputs";
+import { RunnableConfig } from "@langchain/core/dist/runnables/config";
+import { ChainValues } from "@langchain/core/dist/utils/types";
+import { ToolMessage } from "@langchain/core/messages";
+import {
+  currentLogger,
+  currentSpan,
+  Logger,
+  NOOP_SPAN,
+  Span,
+  StartSpanArgs,
+} from "braintrust";
+
+const EXCLUDE_METADATA_PROPS = /^(l[sc]_|langgraph_|__pregel_|checkpoint_ns)/;
+
+/**
+ * A Braintrust tracer for LangChain.js that logs LLM calls, chains, and tools
+ */
+export class BraintrustCallbackHandler<IsAsyncFlush extends boolean = false>
+  extends BaseCallbackHandler
+  implements BaseCallbackHandlerInput
+{
+  name = "BraintrustCallbackHandler";
+  private spans: Map<string, Span>;
+  private logger: Logger<IsAsyncFlush>;
+
+  constructor(logger?: Logger<IsAsyncFlush>) {
+    super();
+
+    this.spans = new Map();
+
+    logger = logger ?? currentLogger();
+    if (!logger) {
+      throw new Error("No logger provided or available.");
+    }
+
+    this.logger = logger;
+  }
+
+  protected startSpan({
+    runId,
+    parentRunId,
+    ...args
+  }: StartSpanArgs & {
+    runId: string;
+    parentRunId?: string;
+  }) {
+    if (this.spans.has(runId)) {
+      // XXX: see graph test case of an example where this _may_ be intended
+      console.warn(
+        `Span already exists for runId ${runId} (this is likely a bug)`,
+      );
+      return;
+    }
+
+    args.event = {
+      ...args.event,
+      metadata: {
+        ...args.event?.metadata,
+        runId,
```

Comment:
I _think_ this is not useful data for users. May need to conditionally include this (perhaps only for testing purposes).